doc/mcxrand.1
.\" Copyright (c) 2011 Stijn van Dongen
.TH "mcxrand" 1 "21 Oct 2011" "mcxrand 11-294" "USER COMMANDS "
.po 2m
.de ZI
.\" Zoem Indent/Itemize macro I.
.br
'in +\\$1
.nr xa 0
.nr xa -\\$1
.nr xb \\$1
.nr xb -\\w'\\$2'
\h'|\\n(xau'\\$2\h'\\n(xbu'\\
..
.de ZJ
.br
.\" Zoem Indent/Itemize macro II.
'in +\\$1
'in +\\$2
.nr xa 0
.nr xa -\\$2
.nr xa -\\w'\\$3'
.nr xb \\$2
\h'|\\n(xau'\\$3\h'\\n(xbu'\\
..
.if n .ll -2m
.am SH
.ie n .in 4m
.el .in 8m
..
.SH NAME
mcxrand \- random shuffling, removal, addition, and perturbation of edges of graphs
.SH SYNOPSIS

\fBmcxrand\fP [options]

.di ZV
.in 0
.nf \fC
mcxrand -gen K -add N -new-g-mean f    # random graph on K nodes, N edges
mcxrand -imx <name> -remove N -add N   # remove then add edges
mcxrand -imx <name> -shuffle N         # shuffle N edge pairs
mcxrand -imx <name> -noise-radius f    # add noise to add weights
.fi \fR
.in
.di
.ne \n(dnu
.nf \fC
.ZV
.fi \fR

\fBmcxrand\fP
\fB[-imx\fP <fname> (\fIinput matrix\fP)\fB]\fP
\fB[-o\fP <fname> (\fIoutput matrix to <fname>\fP)\fB]\fP
\fB[--write-binary\fP (\fIwrite output in binary format\fP)\fB]\fP
\fB[-gen\fP <num> (\fIgenerate new graph\fP)\fB]\fP
\fB[-remove\fP <num> (\fIremove <num> edges\fP)\fB]\fP
\fB[-add\fP <num> (\fIadd <num> edges not yet present\fP)\fB]\fP
\fB[-shuffle\fP <num> (\fIshuffle edge pair <num> times\fP)\fB]\fP
\fB[-noise-radius\fP <num> (\fImaximum spread to add noise with\fP)\fB]\fP
\fB[-noise-sdev\fP <num> (\fInoise standard deviation\fP)\fB]\fP
\fB[-noise-range\fP <num> (\fInoise range factor\fP)\fB]\fP
\fB[-edge-min\fP <num> (\fIabsolute edge weight minimum\fP)\fB]\fP
\fB[-edge-max\fP <num> (\fIabsolute edge weight maximum\fP)\fB]\fP
\fB[-new-g-mean\fP <num> (\fImean for new edges (Gaussian)\fP)\fB]\fP
\fB[-new-g-sdev\fP <num> (\fIstandard deviation for new edges (Gaussian)\fP)\fB]\fP
\fB[-new-g-radius\fP <num> (\fImaximum spread for new edges (Gaussian)\fP)\fB]\fP
\fB[-new-g-min\fP <num> (\fIlower bound selection (Gaussian)\fP)\fB]\fP
\fB[-new-g-max\fP <num> (\fIupper bound selection (Gaussian)\fP)\fB]\fP
\fB[-skew\fP <num> (\fIskew towards min or max\fP)\fB]\fP
\fB[-h\fP (\fIprint synopsis, exit\fP)\fB]\fP
\fB[--apropos\fP (\fIprint synopsis, exit\fP)\fB]\fP
\fB[--version\fP (\fIprint version, exit\fP)\fB]\fP
.SH DESCRIPTION

This utility is a recent addition to the mcl suite and the schemes
explained below will likely be evolved, simplified, and extended
with future releases\&.

The input graph/matrix, if specified with the \fB-imx\fP option,
has to be in mcl matrix/graph format\&.
You can use label input instead by preprocessing the label input
with \fBmcxload(1)\fP, i\&.e\&.

.di ZV
.in 0
.nf \fC
   mcxload -abc <label-file> | mcxrand [options]
.fi \fR
.in
.di
.ne \n(dnu
.nf \fC
.ZV
.fi \fR

Refer to \fBmcxio(5)\fP for a description of these two input formats\&.
By default \fBmcxrand\fP reads from STDIN\&. Change this with the \fB-imx\fP
option\&.

\fBmcxrand\fP can randomly remove and add edges to a graph, or add gaussian noise
to the edge weights of a graph\&. It can also shuffle edge pairs while
preserving the degree sequence of the graph\&.
In \fIremoval mode\fP (\fB-remove\fP\ \&\fI<num>\fP) and in
\fIaddition mode\fP (\fB-add\fP\ \&\fI<num>\fP)
\fBmcxrand\fP enforces arc symmetry by only working with edges
\fIw(i,j)\fP where \fIi < j\fP and symmetrifying the result and adding any
loops that were present in the input graph just before the output stage\&.

In \fIperturbation mode\fP (\fB-noise-radius\fP, with no other mode specified)
the input can be any graph\&.

\fIShuffle mode\fP (\fB-shuffle\fP\ \&\fI<num>\fP) requires an undirected graph
but will only fail when it picks an arc for which the arc in the reverse
direction is not present\&. This means it may or may not fail on directed
input depending on the random choices\&. It does not check equality of the two
arc weights and randomly picks one to represent the edge weight\&.

Edge removal, edge creation, and edge perturbation are applied in this order
if both are specified\&. Edge shuffling presently cannot be combined with one
of the previous modes\&.

A random graph can be generated with \fB-gen\fP\ \&\fIk\fP, which
specifies the number of nodes the graph should have\&. It is equivalent
with pasing (the file name of) an empty graph of the same dimensions as the argument
to \fB-imx\fP\&.

When adding (i\&.e\&. creating) edges, the default is to use the uniform
distribution for new edge weights ranging in some interval\&.
The default interval is [0,1] and can be modified using the
\fB-edge-min\fP\ \&\fImin\fP and \fB-edge-max\fP\ \&\fImax\fP options\&.
A Gaussian edge weight distribution can be obtained by specifying
\fB-new-g-mean\fP\ \&\fInum\fP\&. The standard deviation is by default
1\&.0 and can be altered with \fB-new-g-sdev\fP\ \&\fInum\fP\&.
Currently the edge weigths are generated within the interval
[\fImean-radius\fP, \fImean+radius\fP] where \fIradius\fP is specified
with \fB-new-g-radius\fP\&. They are further selected to
lie within the interval \fI[L,R]\fP if and only if
\fB-new-g-min\fP\ \&\fIL\fP and \fB-new-g-max\fP\ \&\fIR\fP have
been specified\&.

For both uniform and Gaussian edge creation the edge weights
can be skewed towards either side of the distribution with
\fB-skew\fP\ \&\fIc\fP\&. Skewing is applied by mapping
the edge weights to the interval [0,1], applying
the function \fIx^c\fP, and remapping the resulting number\&.
For values \fIc<1\fP this skews the edge weights towards
the right bound and for values \fIc>1\fP this skews the edge
weights towards the left bound\&.
This is a rather crude approach that will likely be
changed in the future\&.

Edge weights can be perturbed by specifying \fB-noise-radius\fP\ \&\fIradius\fP\&.
This sets the maximum perturbation allowed\&. Noise is generated with
a standard deviation that is by default set to 0\&.5 and can be altered
using \fB-noise-sdev\fP\ \&\fInum\fP\&.
Values are generated in the interval \fI[-fac*sdev, fac*sdev]\fP
where \fIfac\fP is set with \fB-noise-range\fP\ \&\fIfac\fP\&.
This interval is mapped to the interval \fI[-radius, radius]\fP before
the resulting value is added to the actual edge weight\&.
This becomes the new value\&. If an interval \fI[L,R]\fP is explicitly
specified using \fB-edge-min\fP\ \&\fIL\fP and \fB-edge-max\fP\ \&\fIR\fP
then the new value will be accepted only if it lies within the interval,
otherwise the edge will not be perturbed\&.
.SH OPTIONS

.ZI 2m "\fB-imx\fP <fname> (\fIinput matrix\fP)"
\&
.br
The file name for input\&. STDIN is assumed if neither
\fB-imx\fP nor \fB-gen\fP\ \&\fInum\fP is specified\&.
.in -2m

.ZI 2m "\fB-o\fP <fname> (\fIoutput matrix to <fname>\fP)"
\&
.br
The file to write the transformed matrix to\&.
.in -2m

.ZI 2m "\fB--write-binary\fP (\fIwrite output in binary format\fP)"
\&
.br
Write the output matrix in native binary format\&.
.in -2m

.ZI 2m "\fB-shuffle\fP <num> (\fIshuffle edge pair <num> times\fP)"
\&
.br
Shuffle edge pair <num> times\&. An edge shuffle
acts on two randomly chosen edges
edges \fIw(a,b)\fP and \fIw(c,d)\fP where all the nodes must

NEWS
$Id: NEWS,v 1.11 2003/11/29 21:30:05 flux Exp $

BLAST users may consider using the blast module. Add --enable-blast to
./configure and read the manual pages of mcxdeblast and mcxassemble.  and
optionally mclblastline. The blast module does not conflict with the tribe
module; you can install both of them simultaneously.

Read ChangeLog.


src/shmx/mcxrand.c
/*   (C) Copyright 2006, 2007, 2008, 2009, 2010, 2011 Stijn van Dongen
 *
 * This file is part of MCL.  You can redistribute and/or modify MCL under the
 * terms of the GNU General Public License; either version 3 of the License or
 * (at your option) any later version.  You should have received a copy of the
 * GPL along with MCL, in the file COPYING.
*/

/*
 * TODO
 *    -  generate random clusterings
 *    -  warn on strange/wrong combinations of options (shuffle & xxx)
 *    -  support DAGs
 *    -  for small matrices and many additions keep track of
 *          matrix with non-edges (worthwhile?).
 *    -  Poisson distribution, others.
 *    #  prove that Pierre's approach to Gaussians works.
 *    -  pick value according to sampled distribution or removals
 *    -  we use only lower bits for edge selection
 *    -  insertion is slow due to array insertion;
 *          ideally store graph in a hash.
*/

#define DEBUG 0

#include <stdio.h>
#include <unistd.h>
#include <string.h>
#include <stdlib.h>
#include <unistd.h>
#include <math.h>
#include <stdlib.h>
#include <stdio.h>
#include <limits.h>
#include <string.h>
#include <ctype.h>
#include <signal.h>
#include <time.h>

#include "impala/io.h"
#include "impala/stream.h"
#include "impala/ivp.h"
#include "impala/app.h"
#include "impala/compose.h"
#include "impala/iface.h"

#include "util/types.h"
#include "util/ding.h"
#include "util/array.h"
#include "util/ting.h"
#include "util/io.h"
#include "util/err.h"
#include "util/equate.h"
#include "util/rand.h"
#include "util/opt.h"

#include "mcl/proc.h"
#include "mcl/procinit.h"
#include "mcl/alg.h"

#include "clew/clm.h"
#include "gryphon/path.h"


const char* me = "mcxrand";

void usage
(  const char**
)  ;


enum
{  MY_OPT_APROPOS
,  MY_OPT_HELP
,  MY_OPT_VERSION
,  MY_OPT_OUT
,  MY_OPT_WB
,  MY_OPT_PLUS
,  MY_OPT_IMX
,  MY_OPT_GEN
,  MY_OPT_ADD
,  MY_OPT_N_SDEV
,  MY_OPT_SHUFFLE
,  MY_OPT_REMOVE
,  MY_OPT_N_RADIUS
,  MY_OPT_N_RANGE
,  MY_OPT_E_MIN
,  MY_OPT_E_MAX
,  MY_OPT_G_MEAN
,  MY_OPT_G_SDEV
,  MY_OPT_G_RADIUS
,  MY_OPT_G_MIN
,  MY_OPT_G_MAX
,  MY_OPT_SKEW
}  ;

const char* syntax = "Usage: mcxrand [options] -imx <mx-file>";

mcxOptAnchor options[] =
{
   {  "--help"
   ,  MCX_OPT_DEFAULT | MCX_OPT_INFO
   ,  MY_OPT_APROPOS
   ,  NULL
   ,  "print this help"
   }
,  {  "-h"
   ,  MCX_OPT_DEFAULT | MCX_OPT_INFO
   ,  MY_OPT_HELP
   ,  NULL
   ,  "print this help"
   }
,  {  "--version"
   ,  MCX_OPT_DEFAULT | MCX_OPT_INFO
   ,  MY_OPT_VERSION
   ,  NULL
   ,  "print version information"
   }
,  {  "-gen"
   ,  MCX_OPT_HASARG
   ,  MY_OPT_GEN
   ,  "<num>"
   ,  "node count"
   }
,  {  "-imx"
   ,  MCX_OPT_HASARG
   ,  MY_OPT_IMX
   ,  "<fname>"
   ,  "input matrix"
   }
,  {  "-noise-radius"
   ,  MCX_OPT_HASARG
   ,  MY_OPT_N_RADIUS
   ,  "<num>"
   ,  "edge perturbation maximum absolute size"
   }
,  {  "-noise-sdev"
   ,  MCX_OPT_HASARG
   ,  MY_OPT_N_SDEV
   ,  "<num>"
   ,  "edge perturbation standard deviation"
   }
,  {  "-noise-range"
   ,  MCX_OPT_HASARG
   ,  MY_OPT_N_RANGE
   ,  "<num>"
   ,  "number of standard deviations allowed"
   }
,  {  "-new-g-radius"
   ,  MCX_OPT_HASARG
   ,  MY_OPT_G_RADIUS
   ,  "<num>"
   ,  "maximum spread to generate new edges with"
   }
,  {  "-new-g-mean"
   ,  MCX_OPT_HASARG
   ,  MY_OPT_G_MEAN
   ,  "<num>"
   ,  "mean to generate new edges with"
   }
,  {  "-new-g-min"
   ,  MCX_OPT_HASARG
   ,  MY_OPT_G_MIN
   ,  "<num>"
   ,  "absolute lower bound for generated weights"
   }
,  {  "-new-g-max"
   ,  MCX_OPT_HASARG
   ,  MY_OPT_G_MAX
   ,  "<num>"
   ,  "absolute upper bound for generated weights"
   }
,  {  "-new-g-sdev"
   ,  MCX_OPT_HASARG
   ,  MY_OPT_G_SDEV
   ,  "<num>"
   ,  "standard deviation to generate new edges with"
   }
,  {  "-edge-min"
   ,  MCX_OPT_HASARG
   ,  MY_OPT_E_MIN
   ,  "<num>"
   ,  "edge addition weight minimum (use -edge-max as well)"
   }
,  {  "-edge-max"
   ,  MCX_OPT_HASARG
   ,  MY_OPT_E_MAX
   ,  "<num>"
   ,  "edge addition weight maximum (use -edge-min as well)"
   }
,  {  "-skew"
   ,  MCX_OPT_HASARG
   ,  MY_OPT_SKEW
   ,  "<num>"
   ,  "skew towards min (<num> > 1) or max (<num> < 1)"
   }
,  {  "-remove"
   ,  MCX_OPT_HASARG
   ,  MY_OPT_REMOVE
   ,  "<num>"
   ,  "remove <num> edges"
   }
,  {  "--write-binary"
   ,  MCX_OPT_DEFAULT
   ,  MY_OPT_WB
   ,  NULL
   ,  "write binary format"
   }
,  {  "+"
   ,  MCX_OPT_HIDDEN
   ,  MY_OPT_PLUS
   ,  NULL
   ,  "write binary format"
   }
,  {  "-o"
   ,  MCX_OPT_HASARG
   ,  MY_OPT_OUT
   ,  "<fname>"
   ,  "output matrix to <fname>"
   }
,  {  "-add"
   ,  MCX_OPT_HASARG
   ,  MY_OPT_ADD
   ,  "<num>"
   ,  "add <num> edges not yet present"
   }
,  {  "-shuffle"
   ,  MCX_OPT_HASARG
   ,  MY_OPT_SHUFFLE
   ,  "<num>"
   ,  "shuffle edge, repeat <num> times"
   }
,  {  NULL, 0, 0, NULL, NULL }
}  ;


int dimCmp
(  const void*   xp
,  const void*   yp
)
   {  return *((dim*) xp) < *((dim*) yp) ? -1 : *((dim*) xp) > *((dim*) yp) ? 1 : 0
;  }


static void mx_readd_diagonal
(  mclx*  mx
,  mclv*  diag
)
   {  dim i
   ;  for (i=0;i<diag->n_ivps;i++)
      {  mclp* ivp = diag->ivps+i
      ;  if (ivp->val)
         mclvInsertIdx(mx->cols+i, ivp->idx, ivp->val)
   ;  }
   }


void static do_the_shuffle
(  mclx* mx
,  dim N_shuffle
,  dim* offsets   /* size N_COLS(mx) */
,  dim N_edge
,  dim random_ignore
)
   {  dim n_shuffle = 0
   ;  while (n_shuffle < N_shuffle)
      {  unsigned long rx = (unsigned long) random()
      ;  unsigned long ry = (unsigned long) random()
      ;  mclp* ivpll, *ivplr, *ivprl, *ivprr
      ;  dim edge_x, edge_y, *edge_px, *edge_py
      ;  ofs xro, yro, xlo, ylo = -1, vxo, vyo
      ;  long xl, xr, yl, yr
      ;  mclv* vecxl, *vecxr, *vecyl, *vecyr
      ;  double xlval, xrval, ylval, yrval

      ;  if (rx >= random_ignore || ry >= random_ignore)
         continue

      ;  edge_x = rx % N_edge    /* fixme probably not optimal */
      ;  edge_y = ry % N_edge    /* fixme probably not optimal */

      ;  if (!(edge_px = mcxBsearchFloor(&edge_x, offsets, N_COLS(mx), sizeof edge_x, dimCmp)))
         mcxDie(1, me, "edge %ld not found (max %ld)", (long) edge_x, (long) N_edge)

      ;  if (!(edge_py = mcxBsearchFloor(&edge_y, offsets, N_COLS(mx), sizeof edge_y, dimCmp)))
         mcxDie(1, me, "edge %ld not found (max %ld)", (long) edge_y, (long) N_edge)

      ;  vxo   =  edge_px - offsets
      ;  xl    =  mx->dom_cols->ivps[vxo].idx
      ;  vecxl =  mx->cols+vxo
      ;  xro   =  edge_x - offsets[vxo]

      ;  vyo   =  edge_py - offsets
      ;  yl    =  mx->dom_cols->ivps[vyo].idx
      ;  vecyl =  mx->cols+vyo
      ;  yro   =  edge_y - offsets[vyo]

                        /* Offset computation gone haywire */
      ;  if (xro >= vecxl->n_ivps || yro >= vecyl->n_ivps)     /* note: mixed sign comparison */
         mcxDie(1, me, "paradox 1 in %ld or %ld", xl, yl)

      ;  xr = vecxl->ivps[xro].idx
      ;  yr = vecyl->ivps[yro].idx
      ;  xrval = vecxl->ivps[xro].val
      ;  yrval = vecyl->ivps[yro].val

                        /* Impossible, should have graph */
      ;  vecxr = mclxGetVector(mx, xr, EXIT_ON_FAIL, NULL)
      ;  vecyr = mclxGetVector(mx, yr, EXIT_ON_FAIL, NULL)

                        /* check that we have four different nodes
                         * loops are not present so no need to check those
                        */
      ;  if (xl == yl || xl == yr || xr == yl || xr == yr)
         continue

      ;  if
         (  (0 > (xlo = mclvGetIvpOffset(vecxr, xl, -1)))
         || (0 > (ylo = mclvGetIvpOffset(vecyr, yl, -1)))
         )
         mcxDie
         (  1
         ,  me
         ,  "symmetry violation 1"
            " %ld not found in %ld/%ld OR %ld not found in %ld/%ld"
         ,  (long) xl, (long) vecxr->vid, (long) vecxr->n_ivps
         ,  (long) yl, (long) vecyr->vid, (long) vecyr->n_ivps
         )

                        /* Now:  xl yl :  ivpll
                         *       xl yr :  ivplr
                         *       xr yl :  ivprl
                         *       xr yr :  ivprr
                        */
      ;  xlval = vecxr->ivps[xlo].val
      ;  ylval = vecyr->ivps[ylo].val

      ;  ivpll = mclvGetIvp(vecxl, yl, NULL)
      ;  ivplr = mclvGetIvp(vecxl, yr, NULL)
      ;  ivprl = mclvGetIvp(vecxr, yl, NULL)
      ;  ivprr = mclvGetIvp(vecxr, yr, NULL)

      ;  if
         (  (ivpll && !mclvGetIvp(vecyl, xl, NULL))
         || (ivplr && !mclvGetIvp(vecyr, xl, NULL))
         || (ivprl && !mclvGetIvp(vecyl, xr, NULL))
         || (ivprr && !mclvGetIvp(vecyr, xr, NULL))
         )
         mcxDie(1, me, "symmetry violation 2")

      ;  if ((ivpll && ivplr) || (ivprl && ivprr))
         continue

      ;  {  if (!ivpll && !ivprr) 
            {        /* vecxl <-> xr   becomes vecxl <-> yl
                      * vecxr <-> xl   becomes vecxr <-> yr
                      * vecyl <-> yr   becomes vecyl <-> xl 
                      * vecyr <-> yl   becomes vecyr <-> xr
                     */
            ;  if
               (  mclvReplaceIdx(vecxl, xro, yl, xrval)
               || mclvReplaceIdx(vecyl, yro, xl, xrval)
               || mclvReplaceIdx(vecxr, xlo, yr, ylval)
               || mclvReplaceIdx(vecyr, ylo, xr, ylval)
               )
               mcxDie(1, me, "parallel replacement failure\n")
#if DEBUG
;fprintf(stderr, "parallel edge change remove  %d-%d  %d-%d add  %d-%d  %d-%d\n",
   vecxl->vid, xr, vecyr->vid, yl,  vecxl->vid, yl, vecyr->vid, xr)
#endif
         ;  }
            else if (!ivplr && !ivprl)
            {        /* vecxl -> xr   becomes vecxl <-> yr
                      * vecxr -> xl   becomes vecxr <-> yl
                      * vecyl -> yr   becomes vecyl <-> xr 
                      * vecyr -> yl   becomes vecyr <-> xl
                     */
               if
               (  mclvReplaceIdx(vecxl, xro, yr, xrval)
               || mclvReplaceIdx(vecyr, ylo, xl, xlval)
               || mclvReplaceIdx(vecxr, xlo, yl, yrval)
               || mclvReplaceIdx(vecyl, yro, xr, yrval)
               )
               mcxDie(1, me, "cross replacement failure\n")
#if DEBUG
;fprintf(stderr, "cross edge change remove  %d-%d  %d-%d add  %d-%d  %d-%d\n",
   vecxl->vid, xr, vecyl->vid, yr,  vecxl->vid, yr, vecyl->vid, xr)
#endif
         ;  }
         }
         n_shuffle++
   ;  }
   }


static dim do_remove
(  mclx* mx
,  dim N_remove
,  dim* offsets      /* size N_COLS(mx) */
,  dim N_edge
,  dim random_ignore
)
   {  dim n_remove = 0
   ;  while (n_remove < N_remove)
      {  unsigned long r = (unsigned long) random()
      ;  dim e, *ep
      ;  ofs xo, yo
      ;  long x
      ;  mclv* vecx

      ;  if (r >= random_ignore)
         continue

      ;  e = r % N_edge

      ;  if (!(ep = mcxBsearchFloor(&e, offsets, N_COLS(mx), sizeof e, dimCmp)))
         mcxDie(1, me, "edge %ld not found (max %ld)", (long) e, (long) N_edge)

      ;  xo = ep - offsets
;if(DEBUG)fprintf(stderr, "have e=%d o=%d\n", (int) e, (int) xo)
      ;  x = mx->dom_cols->ivps[xo].idx
      ;  vecx = mx->cols+xo;
      ;  yo = e - offsets[xo]

      ;  if (yo >= vecx->n_ivps)    /*  note: mixed sign comparison */
         mcxDie
         (  1
         ,  me
         ,  "vector %ld has not enough entries: %ld < %ld"
         ,  (long) vecx->vid
         ,  (long) vecx->n_ivps
         ,  (long) yo
         )

      ;  if (!vecx->ivps[yo].val)
         continue

      ;  vecx->ivps[yo].val = 0.0

;if(DEBUG)fprintf(stderr, "remove [%d] %ld %ld\n", (int) n_remove, (long) x, (long) vecx->ivps[yo].idx)
      ;  n_remove++
   ;  }
      return n_remove
;  }


static dim do_add
(  mclx* mx
,  dim N_add
,  dim N_edge
,  double *l_mean
,  double l_radius
,  double l_sdev
,  double l_min
,  double l_max
,  double skew
,  double e_min
,  double e_max
)
   {  dim n_add = 0
   ;  while (n_add < N_add)
      {  unsigned long r = (unsigned long) random()
      ;  unsigned long s = (unsigned long) random()
      ;  long x, y
      ;  double val
      ;  mclp* ivp

      ;  dim xo = r % N_COLS(mx)       /* fixme, modulo is commonly recommended against */
      ;  dim yo = s % N_COLS(mx)

      ;  if (xo > yo)
         {  long zo = xo
         ;  xo = yo
         ;  yo = zo
      ;  }
         else if (xo == yo)            /* never add loops */
         continue

      ;  x = mx->dom_cols->ivps[xo].idx
      ;  y = mx->dom_cols->ivps[yo].idx

      ;  if (N_edge >= N_COLS(mx) * (N_COLS(mx)-1) / 2)
         break

      ;  ivp = mclvGetIvp(mx->cols+xo, y, NULL)

      ;  if (ivp && ivp->val)
         continue

      ;  if (l_mean)
         {  do
            {  val = mcxNormalCut(l_radius, l_sdev)
            ;  if (skew)
               {  val = (l_radius + val) / (2 * l_radius)
                                 /* ^ map (l_radius + val) to lie within [0,1] */
               ;  val = pow(val, skew)
                                 /* skew it */
               ;  val = (val * 2 * l_radius) - l_radius
                                 /* map it back */
            ;  }
               val += l_mean[0]
         ;  }
            while (l_min < l_max && (val < l_min || val > l_max))
      ;  }
               /* docme: uniform */
         else
         {  val = (((unsigned long) random()) * 1.0) / RAND_MAX
         ;  if (skew)
            val = pow(val, skew)
         ;  val = e_min + val * (e_max - e_min)
      ;  }

         if (!val)
         continue

;if(DEBUG)fprintf(stderr, "add [%d] %ld %ld value %f\n", (int) n_add, (long) x, (long) y, val)
      ;  mclvInsertIdx(mx->cols+xo, y, val)
      ;  N_edge++
      ;  n_add++
   ;  }
      return n_add
;  }


int main
(  int                  argc
,  const char*          argv[]
)  
   {  mcxIO* xfmx          =  mcxIOnew("-", "r"), *xfout = mcxIOnew("-", "w")
   ;  mclx* mx             =  NULL
   ;  mclv* mx_diag        =  NULL

   ;  mcxstatus parseStatus = STATUS_OK
   ;  mcxOption* opts, *opt
   ;  dim N_edge = 0
   ;  dim*  offsets

   ;  dim template_n_nodes = 0
   ;  mcxbool plus = FALSE

   ;  double e_min  = 1.0
   ;  double e_max  = 0.0
   ;  double skew = 0.0

   ;  double radius = 0.0
   ;  double n_sdev  = 0.5
   ;  double n_range = 2.0

   ;  double g_radius  = 0.0
   ;  double g_mean   = 0.0
   ;  double g_sdev = 0.0
   ;  double g_min  = 1.0
   ;  double g_max  = 0.0
   ;  mcxbool do_gaussian = FALSE

   ;  dim i = 0

   ;  dim N_remove  = 0
   ;  dim N_add     = 0
   ;  dim N_shuffle = 0

   ;  unsigned long random_ignore = 0

   ;  srandom(mcxSeed(2308947))
   ;  mcxOptAnchorSortById(options, sizeof(options)/sizeof(mcxOptAnchor) -1)

   ;  if
      (!(opts = mcxOptParse(options, (char**) argv, argc, 1, 0, &parseStatus)))
      exit(0)

   ;  mcxLogLevel =
      MCX_LOG_AGGR | MCX_LOG_MODULE | MCX_LOG_IO | MCX_LOG_GAUGE | MCX_LOG_WARN
   ;  mclxIOsetQMode("MCLXIOVERBOSITY", MCL_APP_VB_YES)
   ;  mclx_app_init(stderr)

   ;  for (opt=opts;opt->anch;opt++)
      {  mcxOptAnchor* anch = opt->anch

      ;  switch(anch->id)
         {  case MY_OPT_HELP
         :  case MY_OPT_APROPOS
         :  mcxOptApropos(stdout, me, syntax, 20, MCX_OPT_DISPLAY_SKIP, options)
         ;  return 0
         ;

            case MY_OPT_VERSION
         :  app_report_version(me)
         ;  return 0
         ;

            case MY_OPT_SKEW
         :  skew = atof(opt->val)
         ;  break
         ;

            case MY_OPT_GEN
         :  template_n_nodes = atoi(opt->val)
         ;  break
         ;

            case MY_OPT_IMX
         :  mcxIOrenew(xfmx, opt->val, NULL)
         ;  break
         ;

            case MY_OPT_PLUS
         :  case MY_OPT_WB
         :  plus = TRUE
         ;  break
         ;

            case MY_OPT_OUT
         :  mcxIOrenew(xfout, opt->val, NULL)
         ;  break
         ;

            case MY_OPT_E_MAX
         :  if (!strcmp(opt->val, "copy"))
            e_max = -DBL_MAX
         ;  else
            e_max = atof(opt->val)
         ;  break
         ;  

            case MY_OPT_E_MIN
         :  e_min = atof(opt->val)
         ;  break
         ;  

            case MY_OPT_G_MIN
         :  g_min = atof(opt->val)
         ;  break
         ;  

            case MY_OPT_G_MAX
         :  g_max = atof(opt->val)
         ;  break
         ;  

            case MY_OPT_G_SDEV
         :  g_sdev = atof(opt->val)
         ;  break
         ;  

            case MY_OPT_G_MEAN
         :  g_mean = atof(opt->val)
         ;  do_gaussian = TRUE
         ;  break
         ;  

            case MY_OPT_G_RADIUS
         :  g_radius = atof(opt->val)
         ;  break
         ;  

            case MY_OPT_N_RANGE
         :  n_range = atof(opt->val)
         ;  break
         ;  

            case MY_OPT_N_SDEV
         :  n_sdev = atof(opt->val)
         ;  break
         ;  

            case MY_OPT_N_RADIUS
         :  radius = atof(opt->val)
         ;  break
         ;  

            case MY_OPT_SHUFFLE
         :  N_shuffle = atoi(opt->val)
         ;  break
         ;  

            case MY_OPT_ADD
         :  N_add = atoi(opt->val)
         ;  break
         ;  
            
            case MY_OPT_REMOVE
         :  N_remove  = atoi(opt->val)
         ;  break
      ;  }
      }

                                 /* hitting y% in vi tells me the size of this block */
      {  if (template_n_nodes)
         mx =  mclxAllocZero
               (  mclvCanonical(NULL, template_n_nodes, 1.0)
               ,  mclvCanonical(NULL, template_n_nodes, 1.0)
               )
      ;  else
         mx =  mclxReadx
               (  xfmx
               ,  EXIT_ON_FAIL
               ,  MCLX_REQUIRE_GRAPH
               )

      ;  mx_diag = mclxDiagValues(mx, MCL_VECTOR_COMPLETE)

      ;  if (N_shuffle)
         mclxAdjustLoops(mx, mclxLoopCBremove, NULL)
      ;  else
         mclxSelectUpper(mx)
      /* ^ apparently we always work on single arc representation (docme andsoon) */

      ;  offsets = mcxAlloc(sizeof offsets[0] * N_COLS(mx), EXIT_ON_FAIL)

      ;  N_edge = 0
      ;  for (i=0;i<N_COLS(mx);i++)
         {  offsets[i] = N_edge
         ;  N_edge += mx->cols[i].n_ivps
      ;  }

         if (N_edge < N_remove)
         {  mcxErr
            (  me
            ,  "removal count %ld exceeds edge count %ld"
            ,  (long) N_remove
            ,  (long) N_edge
            )
         ;  N_remove = N_edge
      ;  }

         random_ignore = RAND_MAX - (N_edge ? RAND_MAX % N_edge : 0)

      ;  if (RAND_MAX / 2 < N_edge)
         mcxDie(1, me, "graph too large!")

      ;  if (N_shuffle)
         {  do_the_shuffle(mx, N_shuffle, offsets, N_edge, random_ignore)
         ;  mx_readd_diagonal(mx, mx_diag)
         ;  mclxWrite(mx, xfout, MCLXIO_VALUE_GETENV, RETURN_ON_FAIL)
         ;  exit(0)
      ;  }

      ;  if (N_remove)
         {  dim n_remove = do_remove(mx, N_remove, offsets, N_edge, random_ignore)
               /* Need to recompute N_edge and random_ignore.
                * NOTE we work with *upper* matrix; this counts graph edges.
               */
         ;  N_edge = mclxNrofEntries(mx) - n_remove
         ;  random_ignore = RAND_MAX - (RAND_MAX % N_COLS(mx))
      ;  }

         if (g_mean)
         {  if (!g_radius)
            {  if (g_sdev)
               g_radius = 2 * g_sdev
            ;  mcxWarn(me, "set radius to %.5f\n", g_radius)
         ;  }
         }

      ;  if (N_add)
         N_edge +=   do_add
                     (  mx
                     ,  N_add  ,  N_edge
                     ,  do_gaussian ? &g_mean : NULL,  g_radius ,  g_sdev ,  g_min ,  g_max
                     ,  skew
                     ,  e_min ,  e_max
                     )

      ;  if (radius)
         {  for (i=0;i<N_COLS(mx);i++)
            {  mclp* ivp = mx->cols[i].ivps, *ivpmax = ivp + mx->cols[i].n_ivps
;if(DEBUG)fprintf(stderr, "here %d\n", (int) i)
            ;  while (ivp < ivpmax)
               {  double val = ivp->val
               ;  double r = mcxNormalCut(n_range * n_sdev, n_sdev)
               ;  double newval = val + radius * (r / (n_range * n_sdev))

               ;  if (e_min < e_max && newval >= e_min && newval <= e_max)
               ;  ivp->val = newval
               ;  ivp++
            ;  }
            }
         }

         mclxUnary(mx, fltxCopy, NULL)    /* remove zeroes */
      ;  mclxAddTranspose(mx, 0.0)
      ;  mx_readd_diagonal(mx, mx_diag)

      ;  if (plus)
         mclxbWrite(mx, xfout, RETURN_ON_FAIL)
      ;  else
         mclxWrite(mx, xfout, MCLXIO_VALUE_GETENV, RETURN_ON_FAIL)
   ;  }
      return 0
;  }


